interfaces/cython/cantera/examples/reactors/ic_engine.py

```diff
@@ -0,0 +1,266 @@
+gas = ct.Solution(reaction_mechanism)
+
+# define initial state
+gas.TPX = T_inlet, p_inlet, comp_inlet
+r = ct.IdealGasReactor(gas)
+# define inlet state
+gas.TPX = T_inlet, p_inlet, comp_inlet
+inlet = ct.Reservoir(gas)
+# define injector state (gaseous!)
+gas.TPX = T_injector, p_injector, comp_injector
+injector = ct.Reservoir(gas)
+# define outlet pressure (temperature and composition don't matter)
+gas.TPX = T_ambient, p_outlet, comp_ambient
+outlet = ct.Reservoir(gas)
+# define ambient pressure (temperature and composition don't matter)
+gas.TPX = T_ambient, p_ambient, comp_ambient
+ambient_air = ct.Reservoir(gas)
+
+# set up connecting devices
+inlet_valve = ct.Valve(inlet, r)
+injector_mfc = ct.MassFlowController(injector, r)
+outlet_valve = ct.Valve(r, outlet)
+piston = ct.Wall(ambient_air, r)
+
+# convert time to crank angle
+def crank_angle(t):
+    return np.remainder(2 * np.pi * f * t, 4 * np.pi)
+
+# set up IC engine parameters
+V_oT = V_H / (epsilon - 1.)
+A_piston = .25 * np.pi * d_piston ** 2
+stroke = V_H / A_piston
+r.volume = V_oT
+piston.area = A_piston
+def piston_speed(t):
+    return - stroke / 2 * 2 * np.pi * f * np.sin(crank_angle(t))
+piston.set_velocity(piston_speed)
+
+# create a reactor network containing the cylinder
+sim = ct.ReactorNet([r])
+
+# set up output data arrays
+t_sim = sim_n_revolutions / f
+t = (np.arange(sim_n_timesteps) + 1) / sim_n_timesteps * t_sim
+p = np.zeros_like(t)
+V = np.zeros_like(t)
+T = np.zeros_like(t)
+s = np.zeros_like(t)
+m = np.zeros_like(t)
+test = np.zeros_like(t)
+mdot_in = np.zeros_like(t)
+mdot_out = np.zeros_like(t)
+MW = np.zeros_like(t)
+d_W_v_d_t = np.zeros_like(t)
+heat_release_rate = np.zeros_like(t)
+species_X = np.zeros((t.size, gas.n_species))
+
+# set parameters for the automatic time step refinement
+n_last_refinement = -np.inf  # for initialization only
+n_wait_coarsening = 10
+
+# do simulation
+for n1, t_i in enumerate(t):
+    # define opening and closing of valves and injector
+    if (np.mod(crank_angle(t_i) - inlet_open, 4 * np.pi) <
+            np.mod(inlet_close - inlet_open, 4 * np.pi)):
+        inlet_valve.set_valve_coeff(inlet_valve_coeff)
+        test[n1] = 1
+    else:
+        inlet_valve.set_valve_coeff(0)
+    if (np.mod(crank_angle(t_i) - outlet_open, 4 * np.pi) <
+            np.mod(outlet_close - outlet_open, 4 * np.pi)):
+        outlet_valve.set_valve_coeff(outlet_valve_coeff)
+    else:
+        outlet_valve.set_valve_coeff(0)
+    if (np.mod(crank_angle(t_i) - injector_open, 4 * np.pi) <
+            np.mod(injector_close - injector_open, 4 * np.pi)):
+        injector_mfc.set_mass_flow_rate(injector_mass / injector_t_open)
+    else:
+        injector_mfc.set_mass_flow_rate(0)
+
+    # perform time integration, refine time step if necessary
+    for n2 in range(4):
+        if n2 is 4:
+            raise 'Error: Refinement limit reached'
+        try:
+            sim.advance(t_i)
+            continue
+        except:
```

Comment:
Prefer to catch an exception of some type -- even ``Exception`` is fine -- so that you don't unexpectedly catch things such as ``KeyboardInterrupt``